File: Revario-Go/Day12/day12.go
package main

import (
	"fmt"
	"github.com/revario/advent_of_code_2016/Revario-Go/common"
	"strconv"
	"strings"
)

func main() {
	input := `input.txt`
	fmt.Printf(`Part 1: %v, Part 2: %v`, parseInstructions(common.ReadFileByRow(input), false), parseInstructions(common.ReadFileByRow(input), true))

}

func parseInstructions(i []string, part2 bool) int {
	var reg = make(map[string]int)
	if part2 {
		reg[`c`] = 1
	}
	for n := 0; n < len(i); n++ {
		inst := strings.Split(i[n], ` `)
		switch inst[0] {
		case `cpy`:

			if num, err := strconv.Atoi(inst[1]); err == nil {
				reg[inst[2]] = num
			} else {
				reg[inst[2]] = reg[inst[1]]
			}

		case `jnz`:
			if d, err := strconv.Atoi(inst[1]); err == nil {
				if d != 0 {
					if num, err := strconv.Atoi(inst[2]); err == nil {
						n += num - 1
					} else {
						panic("Conversion error")
					}
				}

			} else {
				if reg[inst[1]] != 0 {
					if num, err := strconv.Atoi(inst[2]); err == nil {
						n += num - 1
					} else {
						panic("Conversion error")
					}
				}
			}

		case `inc`:
			reg[inst[1]]++

		case `dec`:
			reg[inst[1]]--
		default:
			panic(inst[0])
		}

	}
	return reg[`a`]
}

File: dlofstrom-python/Day12/day12.py

input = open('day12-input.txt','r')
input = [l.strip('\r\n').split(' ',1) for l in input]

registers = {'a':0, 'b':0, 'c':0, 'd':0}

def cpy(xy):
    #print 'cpy', xy
    x, y = xy.split(' ')
    if x in registers:
        registers[y] = registers[x]
    else:
        registers[y] = int(x)
    return 1

def inc(x):
    #print 'inc', x
    registers[x] += 1
    return 1

def dec(x):
    #print 'dec', x
    registers[x] -= 1
    return 1

def jnz(xy):
    #print 'jnz', xy
    x, y = xy.split(' ')
    if x in registers:
        c = registers[x]
    else:
        c = int(x)
    if not c==0:
        return int(y)
    return 1

def assembunny_eval(reg):
    i = 0
    while True:
        #print i,registers
        i += eval(input[i][0]+'(\''+input[i][1]+'\')')
        if not i in range(len(input)):
            break
    return registers[reg]
print 'Part 1:', assembunny_eval('a')

registers = {'a':0, 'b':0, 'c':1, 'd':0}
print 'Part 2:', assembunny_eval('a')
